### PR TITLE
test_post_object_upload_size_below_minimum_rgw_max_chunk_size: new testcase for size validation bug

### PR DESCRIPTION
**Pull request by robbat2**
If all of the following are true `ERR_TOO_SMALL` is wrongly returned:
- the bufferlist contains multiple items
- the last item has non-zero length
- that length is smaller than the POST Policy

The check should be `(ofs < min_len)` instead of `(len < min_len)`.

This is further confirmed by the fact that the next line does `s->obj_size = ofs`;

The bug was refactored in 2016, but was introduced in Oct 2012, when
this functionality was first added to RGW in commit 7bb3504d3f0974e9863f536e9af0ce8889d6888f.

Reference: https://github.com/ceph/ceph/blob/933a42f9af349b3b222270e7f19f1fe151d89e8e/src/rgw/rgw_op.cc#L4474-L4513
Reference: https://github.com/ceph/ceph/commit/7bb3504d3f0974e9863f536e9af0ce8889d6888f
Signed-off-by: Robin H. Johnson <rjohnson@digitalocean.com>

### CODE REVIEW

File: s3tests_boto3/functional/test_s3.py [hunk 1]
```diff
@@ -2817,6 +2817,53 @@ def test_post_object_upload_size_below_minimum():
...
+def test_post_object_upload_size_rgw_chunk_size_bug():
+    # TODO: if this value is different in Teuthology runs, this would need tuning
+    _rgw_max_chunk_size = 4 * 2**20 # 4MiB
+    min_size = _rgw_max_chunk_size * 2
+    max_size = _rgw_max_chunk_size * 3
+    # [(chunk),(small)]
+    test_payload_size = min_size * 2 + 200 # extra bit to push it over the chunk boundary
+    # it should be valid when we run this test!
+    assert test_payload_size > min_size
+    assert test_payload_size < max_size
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2831]: double-checking my math here..
```
rgw_max_chunk_size = 4M
min_size = 8M
max_size = 12M
test_payload_size = 16M + 200
```
so won't this `max_size` assertion fail?

in the interest of a minimal reproducer, could we just use `test_payload_size = _rgw_max_chunk_size + 200`?

File: s3tests_boto3/functional/test_s3.py [hunk 2]
```diff
@@ -2817,6 +2817,53 @@ def test_post_object_upload_size_below_minimum():
 
     r = requests.post(url, files=payload, verify=get_config_ssl_verify())
     assert r.status_code == 400
+    assert r.error_code == 'EntityTooSmall'
+
+def test_post_object_upload_size_rgw_chunk_size_bug():
+    # TODO: if this value is different in Teuthology runs, this would need tuning
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2823]: the rgw suite does vary the *stripe* size, but the chunk size stays constant at 4M: https://github.com/ceph/ceph/blob/main/qa/suites/rgw/verify/striping%24/stripe-greater-than-chunk.yaml

File: s3tests_boto3/functional/test_s3.py [hunk 3]
```diff
@@ -2817,6 +2817,53 @@ def test_post_object_upload_size_below_minimum():
...
+    _rgw_max_chunk_size = 4 * 2**20 # 4MiB
+    min_size = _rgw_max_chunk_size * 2
+    max_size = _rgw_max_chunk_size * 3
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2826]: we try to keep rgw specifics out s3-tests, but i'm happy with the way you've done this 👍 it exercises a corner case for rgw, but should succeed regardless of the server. since it mentions an rgw bug, please add a comment that links to https://tracker.ceph.com/issues/58627

File: s3tests_boto3/functional/test_s3.py (hunk 1 above)
Comment (robbat2) [reply s3tests_boto3/functional/test_s3.py:2831]: oops, editing fail when I cleaned this from internal version

Comment (robbat2) [reply s3tests_boto3/functional/test_s3.py]: Fixed

File: s3tests_boto3/functional/test_s3.py (hunk 3 above)
Comment (robbat2) [reply s3tests_boto3/functional/test_s3.py:2828]: Done

File: s3tests_boto3/functional/test_s3.py (hunk 2 above)
Comment (robbat2) [reply s3tests_boto3/functional/test_s3.py:2824]: Thanks, linked

File: s3tests_boto3/functional/test_s3.py [hunk 4]
```diff
@@ -2817,6 +2817,55 @@ def test_post_object_upload_size_below_minimum():
...
+
+    signature = base64.b64encode(hmac.new(bytes(aws_secret_access_key, 'utf-8'), policy, hashlib.sha1).digest())
+
+    payload = OrderedDict([ ("key" , "foo.txt"),("AWSAccessKeyId" , aws_access_key_id),\
+    ("acl" , "private"),("signature" , signature),("policy" , policy),\
+    ("Content-Type" , "text/plain"),('file', (test_payload))])
+
+    r = requests.post(url, files=payload, verify=get_config_ssl_verify())
+    assert r.status_code == 400
+    assert r.error_code == 'EntityTooSmall'
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2868]: shouldn't we expect success in this case?

Comment (cbodley) [reply s3tests_boto3/functional/test_s3.py:2868]: ping @robbat2

Comment (robbat2) [reply s3tests_boto3/functional/test_s3.py:2868]: Good catch, can you please amend and merge? DM me for $reasons.

Comment (cbodley) [reply s3tests_boto3/functional/test_s3.py:2868]: sure. no explanation necessary

**Review by cbodley — APPROVED** [review]

File: s3tests_boto3/functional/test_s3.py [hunk 5]
```diff
@@ -2817,6 +2817,54 @@ def test_post_object_upload_size_below_minimum():
 
     r = requests.post(url, files=payload, verify=get_config_ssl_verify())
     assert r.status_code == 400
+    assert r.error_code == 'EntityTooSmall'
```
Comment (cbodley) [inline s3tests_boto3/functional/test_s3.py:2820]: this fails with `AttributeError: 'Response' object has no attribute 'error_code'`

this is a call to `requests.post()` which doesn't understand these s3 error codes. i think we'd have to parse the xml response to get at it. i'm going to remove this commit for now
